Rewrite the shared JsonSerializerOptions cache implementation.

eiriktsarpalis opened the pull request:
Rewrites the shared `JsonSerializerOptions.CachingContext` implementation so that matching instances are looked up using lock-free linear search over an array of `WeakReference<CachingContext>` instances. This achieves a couple of things:

1. Fixes #76548, in which user-defined converters referencing a large amount memory could end up being rooted in the global cache in scenaria where serialization operations are not performed frequently.
2. Improves performance compared to the existing `ConcurrentDictionary` approach:

|                   Method |        Job |                                                                                                          Branch |        Mean |       Error |      StdDev |      Median |         Min |         Max | Ratio | MannWhitney(3%) | RatioSD |  Gen 0 |  Gen 1 |  Gen 2 | Allocated | Alloc Ratio |
|------------------------- |----------- |------------------------------------------------------------------------------------------------------------------- |------------:|------------:|------------:|------------:|------------:|------------:|------:|---------------- |--------:|-------:|-------:|-------:|----------:|------------:|
|                          |            |                                                                                                                    |             |             |             |             |             |             |       |                 |         |        |        |        |           |             |
|        NewDefaultOptions | Job-CLWPWY |          main |    998.3 ns |    47.26 ns |    54.43 ns |  1,010.4 ns |    885.2 ns |  1,087.4 ns |  1.00 |            Base |    0.00 | 0.0408 | 0.0041 | 0.0041 |     461 B |        1.00 |
|        NewDefaultOptions | Job-AKKANL | PR |    901.2 ns |    46.17 ns |    53.17 ns |    894.7 ns |    832.5 ns |  1,001.2 ns |  0.91 |          Faster |    0.08 | 0.0398 | 0.0040 | 0.0040 |     461 B |        1.00 |
|                          |            |                                                                                                                    |             |             |             |             |             |             |       |                 |         |        |        |        |           |             |
|     NewCustomizedOptions | Job-CLWPWY |          main |  1,079.3 ns |    52.27 ns |    55.93 ns |  1,077.0 ns |    990.1 ns |  1,211.8 ns |  1.00 |            Base |    0.00 | 0.0435 | 0.0040 | 0.0040 |     485 B |        1.00 |
|     NewCustomizedOptions | Job-AKKANL | PR |  1,027.9 ns |    45.85 ns |    52.80 ns |  1,032.2 ns |    938.0 ns |  1,102.2 ns |  0.95 |            Same |    0.07 | 0.0427 | 0.0039 | 0.0039 |     487 B |        1.00 |
|                          |            |                                                                                                                    |             |             |             |             |             |             |       |                 |         |        |        |        |           |             |
|       NewCustomConverter | Job-CLWPWY |          main | 21,462.8 ns | 1,790.02 ns | 1,915.31 ns | 20,926.3 ns | 18,933.0 ns | 25,316.8 ns |  1.00 |            Base |    0.00 | 0.7659 | 0.0766 |      - |    8027 B |        1.00 |
|       NewCustomConverter | Job-AKKANL | PR | 20,540.2 ns |   848.39 ns |   977.01 ns | 20,651.8 ns | 19,212.5 ns | 22,677.2 ns |  0.95 |            Same |    0.09 | 0.7716 | 0.0772 |      - |    8005 B |        1.00 |
|                          |            |                                                                                                                    |             |             |             |             |             |             |       |                 |         |        |        |        |           |             |
| NewCachedCustomConverter | Job-CLWPWY |          main |  1,228.5 ns |    89.28 ns |    99.23 ns |  1,204.1 ns |  1,067.9 ns |  1,422.1 ns |  1.00 |            Base |    0.00 | 0.0502 | 0.0046 | 0.0046 |     545 B |        1.00 |
| NewCachedCustomConverter | Job-AKKANL | PR |  1,089.1 ns |    43.82 ns |    50.47 ns |  1,105.0 ns |    982.9 ns |  1,159.6 ns |  0.89 |          Faster |    0.08 | 0.0509 | 0.0039 | 0.0039 |     545 B |        1.00 |

eiriktsarpalis committed:
Rewrite the shared JsonSerializerOptions cache implementation.

eiriktsarpalis commented:
FYI @SteveSandersonMS

stephentoub commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:
```diff
@@ -163,146 +162,124 @@ public void Clear()
-                    }
-                }
+                WeakReference<CachingContext>?[] trackedContexts = s_trackedContexts;
+                int size = s_size;
```
This is accessing shared mutable state without a lock.  At a minimum we should have a comment explaining why that's ok.

stephentoub commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:
```diff
@@ -163,146 +162,124 @@ public void Clear()
+                for (int i = nextAvailable; i < size; i++)
+                {
+                    trackedOptions[i] = null;
                 }
```
```C#
Array.Clear(trackedOptions, nextAvailable, size - nextAvailable);
```
?

stephentoub commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:
```diff
@@ -163,146 +162,124 @@ public void Clear()
             private const int MaxTrackedContexts = 64;
-            private static readonly ConcurrentDictionary<JsonSerializerOptions, WeakReference<CachingContext>> s_cache =
-                new(concurrencyLevel: 1, capacity: MaxTrackedContexts, new EqualityComparer());
+            private static readonly WeakReference<CachingContext>?[] s_trackedContexts = new WeakReference<CachingContext>[MaxTrackedContexts];
```
Entries in the cache could end up being removed even on a gen0 GC if there's no rooted JsonSerializerOptions currently referencing that CachingContext.  How likely is that?

eiriktsarpalis replied:
It is likely, although this is also true for the current implementation which also stores the cache inside a WeakReference.

jkotas replied:
Yes, this cache works nice in micro-benchmarks and small apps. It is unlikely to be effective in larger apps that do number of different things where the GC is going to be likely to clear the weak references too often.

eiriktsarpalis replied:
We could replace the WeakReferences altogether with a manual eviction mechanism with an explicit TTL, although this will most likely involve setting up a `Timer` in the background. I think we discussed doing that last year but I couldn't find any prior art of other libraries doing anything similar.

jkotas replied:
The timer just comes with different set of problems. I do not think that it makes sense to magically fix the problem for everybody.

Instead, we should spend energy on making sure that people do their own caching correctly and avoid this problem all together.

stephentoub replied:
We should have an analyzer that flags these kinds of cases, in particular constructing a new JsonSerializerOptions for each Serialize call, especially if its only populated with constants / always the same way.

We still need a cache, though, even if it doesn't help all cases.  The cliff is too steep otherwise.

eiriktsarpalis replied:
Related: https://github.com/dotnet/runtime/issues/65396

eiriktsarpalis replied:
> I do not think that it makes sense to try to magically fix the problem for everybody.

It certainly cannot fix the problem for all use cases. The equality comparison used to determine configuration equivalence is pretty conservative. For example, it only compares custom converters up to reference equality, which is unlikely to happen in cases where custom converters are being configured.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:
```diff
@@ -163,146 +162,124 @@ public void Clear()
+                }
 
-                if (cache.TryGetValue(options, out WeakReference<CachingContext>? wr) && wr.TryGetTarget(out CachingContext? ctx))
+                if (s_size == MaxTrackedContexts && !foundDanglingEntries)
```
Yes, that is pretty much the intention. I'll add a remark making that more explicit.

eiriktsarpalis committed:
address feedback

jkotas commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:
```diff
@@ -311,9 +287,9 @@ static int EstimateEvictionRunsToSkip(int latestEvictionCount)
         /// the converse however does not necessarily hold.
         /// </summary>
-        private sealed class EqualityComparer : IEqualityComparer<JsonSerializerOptions>
+        private static class EqualityComparer
```
This does not need to be a dedicated class anymore

jkotas commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs:
```diff
@@ -163,146 +162,123 @@ public void Clear()
+                    {
+                        lock (s_trackedContexts)
+                        {
+                            EvictDanglingEntries();
```
Is the whole purpose of this call to just clear the WeakReferences that got cleared?  I do not think you need to bother with that. Cleared WeakReferences cost very little. If the app peaked at certain number of caches entries, it is likely going to peak at the same number again - it is better to keep reusing the WeakReferences.

eiriktsarpalis replied:
It is also meant to push longer surviving caches towards the start of the array, so they can be found more quickly. Perhaps it's not too important though.

eiriktsarpalis committed:
Simplify cache implementation& address feedback.

eiriktsarpalis commented:
I've simplified the cache implementation so that evictions are no longer performed. Instead, every lookup will perform a linear search across all 64 entries until a matching one is found. Updated benchmarks:

|                   Method |        Job |                                                                                                          Branch |        Mean |     Error |    StdDev |      Median |         Min |         Max | Ratio | MannWhitney(3%) | RatioSD |  Gen 0 |  Gen 1 |  Gen 2 | Allocated | Alloc Ratio |
|------------------------- |----------- |------------------------------------------------------------------------------------------------------------------- |------------:|----------:|----------:|------------:|------------:|------------:|------:|---------------- |--------:|-------:|-------:|-------:|----------:|------------:|
|        NewDefaultOptions | Job-FFWUTS |          main |  1,013.0 ns |  43.04 ns |  49.57 ns |  1,007.6 ns |    920.9 ns |  1,102.3 ns |  1.00 |            Base |    0.00 | 0.0405 | 0.0040 | 0.0040 |     461 B |        1.00 |
|        NewDefaultOptions | Job-ZLXSEW | PR |    960.6 ns |  41.54 ns |  47.83 ns |    944.0 ns |    896.3 ns |  1,092.7 ns |  0.95 |            Same |    0.07 | 0.0453 | 0.0038 | 0.0038 |     459 B |        1.00 |
|                          |            |                                                                                                                    |             |           |           |             |             |             |       |                 |         |        |        |        |           |             |
|     NewCustomizedOptions | Job-FFWUTS |          main |  1,141.8 ns |  51.71 ns |  59.55 ns |  1,119.8 ns |  1,068.5 ns |  1,279.2 ns |  1.00 |            Base |    0.00 | 0.0432 | 0.0043 | 0.0043 |     484 B |        1.00 |
|     NewCustomizedOptions | Job-ZLXSEW | PR |  1,056.7 ns |  53.45 ns |  61.55 ns |  1,051.2 ns |    938.2 ns |  1,184.9 ns |  0.93 |          Faster |    0.06 | 0.0424 | 0.0039 | 0.0039 |     486 B |        1.00 |
|                          |            |                                                                                                                    |             |           |           |             |             |             |       |                 |         |        |        |        |           |             |
|       NewCustomConverter | Job-FFWUTS |          main | 20,046.4 ns | 439.57 ns | 506.20 ns | 20,003.1 ns | 19,403.1 ns | 20,968.4 ns |  1.00 |            Base |    0.00 | 0.7512 | 0.0751 |      - |    8026 B |        1.00 |
|       NewCustomConverter | Job-ZLXSEW | PR | 20,828.4 ns | 843.85 ns | 971.78 ns | 20,691.9 ns | 19,709.1 ns | 23,267.3 ns |  1.04 |            Same |    0.06 | 0.7166 | 0.0796 |      - |    8004 B |        1.00 |
|                          |            |                                                                                                                    |             |           |           |             |             |             |       |                 |         |        |        |        |           |             |
| NewCachedCustomConverter | Job-FFWUTS |          main |  1,115.7 ns |  39.86 ns |  45.90 ns |  1,124.0 ns |  1,028.3 ns |  1,177.7 ns |  1.00 |            Base |    0.00 | 0.0504 | 0.0046 | 0.0046 |     545 B |        1.00 |
| NewCachedCustomConverter | Job-ZLXSEW | PR |  1,075.1 ns |  34.65 ns |  39.90 ns |  1,071.3 ns |  1,010.3 ns |  1,143.5 ns |  0.97 |            Same |    0.05 | 0.0501 | 0.0046 | 0.0046 |     545 B |        1.00 |

jkotas reviewed (approved):
LGTM

eiriktsarpalis commented:
/backport to release/7.0
